restricted_site_access.php

```diff
@@ -122,6 +122,32 @@ public static function add_actions() {
 
 		add_filter( 'pre_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
 		add_filter( 'pre_site_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
+		add_filter( 'application_password_is_api_request', array( __CLASS__, 'is_api_request' ) );
+	}
+
+	/**
+	 * Add filter to override application_password_is_api_request filter. This has to be done since rest API isn't been loaded until we call restrict_access.
+	 *
+	 * @param bool $original_value Original value passed by filter.
+	 * @return bool
+	 */
+	public static function is_api_request( $original_value ) {
+
+		$protocol = ( isset( $_SERVER['HTTPS'] ) && 'on' === $_SERVER['HTTPS'] ? 'https' : 'http' );
+		$request  = isset( $_SERVER['HTTP_HOST'] ) && sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) )
+		. isset( $_SERVER['REQUEST_URI'] ) && sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) );
+
+		// We need to determine current URL via php because $wp->request is not set at this point.
+		$current_request = esc_url_raw( "$protocol://$request" );
```

Comment:
While testing this, I was seeing `$current_request` as `http://1`. I did my testing by dumping `header( "X-RSA-Request: $current_request" )` in the code. I've split it out a little to get it working:

Stripping the query string is needed for the next item.

```suggestion
		$protocol = ( isset( $_SERVER['HTTPS'] ) && 'on' === $_SERVER['HTTPS'] ) ? 'https' : 'http';
		$host     = isset( $_SERVER['HTTP_HOST'] ) ? sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) : '';
		$path     = isset( $_SERVER['REQUEST_URI'] ) ? sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ) : '';

		// We need to determine current URL via php because $wp->request is not set at this point.
		$current_request = esc_url_raw( "{$protocol}://{$host}{$path}" );
		// Strip the query string.
		$current_request = explode( '?', $current_request, 2 )[0];
```